Collection of Makefile housekeeping

rwe opened the pull request:
This includes a bunch of teensy improvements, nothing huge. `make && make test` had lots of strange failures for me when run out of the box with a normal stack configuration.

Some highlights:
  * Set safe-ish bash flags (`set -Eeu -o pipefail`) for make recipes. This should prevent (or at least reduce) make barging head when things go wrong.
  * Consistently invoke `ghc` or `runghc` via `stack ghc`/`stack runghc`, if using stack. This solved a lot of head-scratching issues for me.
  * Along the same lines, don't force `--system-ghc` for stack. The user can specify that, but they probably don't want it. I've left it enabled for Travis just in case.
  * No longer depend on systems having `wdiff` or `colordiff` installed. (macOS doesn't by default). Instead it uses `git diff`, which will use diff coloring if the user has that set up, and they probably do. Also it's guaranteed to be installed.

This cleans up a few trivial consistency issues and one or two makefile bugs caught by the stricter bash flags.

rwe committed:
make: Create TMPDIR (e.g. highlighting-tmp) with mkdir -p

If the directory already exists for some reason, this avoids failure.

rwe committed:
make: test/interaction: Use git diff for wdiff | colordiff

Developing on macOS I didn't have wdiff (word diff) or colordiff
installed, so `make test` was puking. However, every system running
tests will at least have `git` installed, so instead using `git diff
--no-index -w` provides the same functionality with fewer external
requirements.

rwe committed:
make: Capitalize SED tool var per convention

Similar to other utilities like $(AWK) or $(CC).
Also made that definition strict.

rwe committed:
make: Use $(SED) variable consistently in makefiles

rwe committed:
make: Add comment on Agda version increment script

rwe committed:
make: Move GHC_VERSION into new mk/ghc.mk

rwe committed:
make: Consistently use runghc instead of runhaskell

All of these instances refer to GHC-only files with special LANGUAGE
pragmas or extra ghc-opts. But other makefiles ran with `runghc`. This
makes it consistent, and avoids confusion if a system ever attempted to
execute with a non-GHC compiler versus giving a command-not-found.

rwe committed:
make: Factor out $(GHC) and $(RUNGHC) commands

rwe committed:
make: Drop _CMD suffix {CABAL,STACK}_CMD

Consistent with other Makefile tool vars like $(CC) and $(AWK).
This also hopefully disambiguates that the variable refers to the
executable (cabal), rather than the verb (install).

rwe committed:
make: Consistently invoke cabal via command variable

rwe committed:
make: Move STACK var under new mk/stack.mk

rwe committed:
make: Consistently invoke stack via $(STACK)

rwe committed:
make: Add HAS_STACK to simplify checking for stack.yaml

rwe committed:
make: Consistently invoke make via $(MAKE)

rwe committed:
make: Only add --system-ghc argument on Travis

Developers running `make` will usually want to use the GHC version
they've specified in their stack.yaml. Otherwise they can specify that
option in in their stack.yaml themselves.

Note that GitHub workflows currently do not use the Makefile, but
instead invoke `stack` directly. (See: .github/workflows/stack.yml)

It's possible that Travis doesn't care either, but seemed better safe
than sorry.

rwe committed:
make: Execute ghc and runghc via stack if using stack

rwe committed:
make: Move Darwin $(sed) definition into mk/common.mk

rwe committed:
make: Add mk/common.mk

This is a quasi-root variables file that checks that `$(TOP)` is defined
and includes the user's `mk/config.mk`, if it exists.

rwe committed:
make: Read PARALLEL_TESTS like any other variable

rwe committed:
make: Use AGDA_BIN a little more consistently

rwe committed:
make: Run initial check diff with -q

The output was discarded anyway. -q suppresses output and returns
immediately if there's a difference.

rwe committed:
make: Fix incorrectly quoted loop in test script

Several quoting issues in this one line:
  for file in ls `$*.*agda`; do touch $file; done ;

Would expand in make to the following shell script (replacing $* with
the make target, and $f replaced an undefined/empty variable):
  for file in ls `Foo.*agda`; do touch ile; done ;

Which bash interprets as: for each word in this list, consisting of the
word "ls", and (whatever comes out of executing a command named
"Foo.*agda"), touch the file named "ile".

This has been updated to just touch any files matching that glob, but
not attempting to create a file if none match.

rwe committed:
make: Set SHELL from mk/common.mk and run with safe flags

andreasabel commented on Makefile:
```diff
@@ -20,18 +20,12 @@ include ./mk/pretty.mk
-else
-# Getting the value from the $(PARALLEL_TESTS_FILE) file.
-include $(PARALLEL_TESTS_FILE)
+# You can use the PARALLEL_TESTS variable to control the number of parallel
```
I wonder whether this file `mk/parallel-tests.mk` wasn't a convenience to define the number of parallel test on your system according to your liking once and for all.

How would you do this now?

rwe replied:
You can add it to `mk/config.mk` along with any other custom overriden variables. That file is loaded if it exists and seemed a better solution than a file per variable.

Or to make this behave exactly as before:
```sh
echo '-include $(TOP)/mk/parallel-tests.mk' >> mk/config.mk
```

`mk/parallel-tests.mk` was undocumented, and scripts did not reference it, so I believe if it's used it's most likely by the people reviewing this 🙂 But I'd be  happy to send a message to the list if it's more widely known.

andreasabel replied:
Local configuration should be in files that are not under version control, as git cannot handle well situations where you make changes to registered files that you never want to push.
If I change `config.mk`, git will annoy me to death.  (`darcs` could deal with commits one never wants to push.)

asr replied:
> But I'd be happy to send a message to the list if it's more widely known.

I know and use it.

rwe replied:
@andreasabel — The path `mk/config.mk` is already `.gitignored`. It does not exist under version control in the same way that `mk/parallel-tests.mk` did not. (`mk/config.mk` was previously never loaded, but now is loaded if it exists).

But I'll restore the loading of `mk/parallel-test.mk` as well since it sounds like it's a concern. It just seemed better to nip the single instance of the "one file per variable" pattern while there was only one case of it.

asr replied:
If `mk/config.mk` is `.gitignored` I have no problem on using it instead of `mk/parallel-tests.mk`.

rwe replied:
I'm doing a couple things to help with this:
  1. ~Changing the name `mk/config.mk` to `mk/custom.mk` to reinforce that this path is `.gitignored`. (Again, the path was not loaded before this PR, so this is cosmetic).~
  1. If `mk/config.mk` does not exist and `mk/parallel-tests.mk` does, emit a small warning + hint and load it. This will prevent any surprising changes for developers. Hopefully that block can be removed in the future.
  2. I'm adding both the existence + purpose of `mk/config.mk` and of the `PARALLEL_TESTS` variable to the `HACKING.md`, so that it's more discoverable.

rwe replied:
I misclick-submitted that comment too early. Although `mk/custom.mk` is a better name, the benefit of `mk/config.mk` is that it was _already_ ignored, so would not show up in branch changes. So scratch that first point.

rwe replied:
That's now pushed

andreasabel replied:
Good, my concern has been addressed!

andreasabel reviewed (commented):
Thanks for this great organizational and solidification work!  I looked through all the patches and only have one question about the `PARALLEL_TESTS_FILE` removal.

gallais commented on test/lib-interaction/Makefile:
```diff
@@ -74,20 +67,20 @@ $(OutFiles) : %.out : $(wildcard %.agda %.lagda) $(wildcard %.in %.in_ghci)
 			cat $*.tmp; \
 			echo "=== Diff ==="; \
-			diff -b $*.out $*.tmp; \
+			( git diff --no-index --no-ext-diff -w $*.out $*.tmp || true ) ; \
```
Are you looking for the `--exit-code` option to `git diff`?

rwe replied:
No, it was a conscious choice, though possibly deserved a comment. Pagers can sometimes propagate exit codes unrelated to the diff status. Since this is specifically for interactive use it's likely a pager or other user-configured filter is involved. For example, if you use `pygmentize`, for a long time that used to exit from an `EPIPE` when `less` ends. Some other pagers and filters can exit with status if you decline to view the whole output. So this status suppression is intentionally more heavy-handed than `--no-exit-code` permits.

(If you're referring to the `--no-ext-diff`, I'm attempting to keep it terminal-based due to the interactive prompt; some people have GUI diff drivers).

rwe committed:
make/doc: Document purpose and use of user-created mk/config.mk

rwe committed:
make/doc: Document PARALLEL_TESTS variable

rwe committed:
make: Load deprecated mk/parallel-test.mk if exists

The Agda makefiles prior to PR #4841 would load `mk/parallel-test.mk`,
which, if existed, would be expected to define the single variable
`PARALLEL_TESTS`. They did not load `mk/config.mk` or expect that to
exist.

It's simpler to maintain a single source of truth for user
configuration, but we don't want to spring changes on developers who had
customized that option.

So: if the new `mk/config.mk` file does not exist, but the old
`mk/parallel-tests.mk` does, load it and gently nudge them toward
putting creating a `mk/config.mk`.

We don't create it automatically because this loading phase may not be
atomic and we don't want to clobber things that aren't build targets.

This commit should be reverted at some point in the mid future, leaving
`mk/config.mk` as the single source of truth.

rwe commented:
@asr @andreasabel — No urgency, but I believe all your concerns have been addressed here, if the solution makes sense to you.
